### ISSUE TITLE
Non competing subscribers on custom channel

### ISSUE THREAD

**Issue by bzbetty** [redux-saga/redux-saga#820]
I making a small game prototype using redux saga.

Originally I was PUTing clock ticks as messages but it made the standard redux messages rather noisy (using the chrome redux plugin) since it's an even that happens up to 60 times a second.  So i tried switching to using a custom channel instead.

eg  yield put({ type: 'TICK' }) vs  yield put(chan, { type: 'TICK' });

So now i've come across a rather large difference between the standard channel and custom ones - the consumers are competing, first one to grab a message gets it and the others miss out.

is there another mechanism i should be using to put large number of multi subscriber messages around without polluting the standard channel?

**Comment by bzbetty** [comment]
believe I figured it out by diving into the source - need to open up multiple channels and use emitters to support subscribers.  example is the eventChannel.

**Comment by Andarist** [comment]
actually im in the middle of adding a possibility to support multiple subscribers into the channels

I have 2 ideas how this can be achieved:
- exposing second type of channel - multicast, which would by default support multiple subscribers
- creating `putAll` effect which would inform all available subscribers instead of just the first one

Which one sounds more interesting to you? Maybe you would find use cases for both? Second one could be more robust (and maybe more confusing because of that) as it would support both types of emiting at once - `put` (single subscriber notified) and `putAll` (all subscribers notified). Where first one would allow only for a latter.

**Comment by bzbetty** [comment]
I think generally you'd want a channel to either be multicast or round robin and not switch between so I think the first option makes the most sense.

As I mentioned in my case the emitter method worked quite nicely in the end.

```
import { delay, takeEvery, takeLatest, eventChannel, END, channel } from 'redux-saga';
import { race, fork, take, call, put, select } from 'redux-saga/effects';

export default function gameLoop() {
  let lastFrameTimeMs: number = 0;
  let timestep: number = 1000 / 60;

  return eventChannel(emitter => {
    var onAnimationFrame = timestamp => {
      // Throttle the frame rate.    
      if (timestamp >= lastFrameTimeMs + timestep) {        
        lastFrameTimeMs = timestamp;
        emitter({ type: 'GAMELOOP_TICK', payload: { timestamp: timestamp } });
      }
      window.requestAnimationFrame(onAnimationFrame);
    };

    window.requestAnimationFrame(onAnimationFrame);

    return () => { }; //todo unsubscribe
  });

}
```

**Comment by Andarist** [comment]
yeah, if you use them later like this numerous times:
```javascript
const chan = gameLoop()
yield takeEvery(chan, worker)
```
Each will cause a new instance of this particular `eventChannel`, so the do not synchronize between each other etc. Ofc you could use some closure here in order to avoid this problem, but still - you would have many channel instances instead of just one.

Anyway - working on the multicast, so stay tuned! :) And also came to the conclusion that strict difference between unicast vs multicast is the way to go, no need to mix them in some unnatural way.

**Comment by ms88privat** [comment]
This addition would be great. I'm also for `multicastChannel` 👍 

With this type of message bus, redux-saga is really well equipped to handle all kind of system architectures.

@Andarist Did you make any progress yet?

**Comment by Andarist** [comment]
A little bit - https://github.com/redux-saga/redux-saga/pull/824 .  Hadnt time yet to finish work on this, will try to push it forward soon, but no hard promises for any dates :/

**Comment by Andarist** [comment]
This thread got a little bit stale and it's also a duplicate of other issues, so I'm closing it. If you need to discuss this further, please just reply back here.

I have started work on v1, gathering all things that need to be done before the release, you can track the progress [here](https://github.com/redux-saga/redux-saga/milestone/1) and the multicast will be part of the release.
